animations/tween-animations.md

```diff
@@ -0,0 +1,726 @@
+  Animation<double> animation;
+
+  initState() {
+    super.initState();
+    controller = new AnimationController(
+        duration: const Duration(milliseconds: 2000), vsync: this);
+    animation = new Tween(begin: 0.0, end: 300.0).animate(controller)
+      [[highlight]]..addStatusListener((state) => print("$state"));[[/highlight]]
+    controller.forward();
+  }
+  //...
+}
+{% endprettify %}
+
+Running this code produces lines like the following:
+
+<!-- skip -->
+{% prettify sh %}
+AnimationStatus.forward
+AnimationStatus.completed
+{% endprettify %}
+
+Next, use `addStatusListener()` to reverse the animation at the
+beginning or the end. This creates a "breathing" effect:
+
+<!-- skip -->
+{% prettify dart %}
+class _LogoAppState extends State<LogoApp> with SingleTickerProviderStateMixin {
+  AnimationController controller;
+  Animation<double> animation;
+
+  initState() {
+    super.initState();
+    controller = new AnimationController(
+        duration: const Duration(milliseconds: 2000), vsync: this);
+    animation = new Tween(begin: 0.0, end: 300.0).animate(controller);
+
+    animation.addStatusListener((status) {
+      if (status == AnimationStatus.completed) {
+        controller.reverse();
+      } else if (status == AnimationStatus.dismissed) {
+        controller.forward();
+      }
+    });
+    controller.forward();
+  }
+  //...
+}
+{% endprettify %}
+
+You can find the source for this example,
+[animate3,](https://raw.githubusercontent.com/InMatrix/animation_tutorial/master/lib/animate3.dart)
+on GitHub.
+
+## Refactoring with AnimatedBuilder
+
+One problem with the code in the previous example, is that changing
+the animation required changing the widget that renders the logo.
+A better solution is to separate responsibilities into different
+classes:
+
+* Render the logo
+* Define the Animation object
+* Render the transition
+
+You can accomplish this separation with the help of the
+AnimatedBuilder class. An AnimatedBuilder is a separate class in the
+render tree. Like AnimatedWidget, AnimatedBuilder automatically
+listens to notifications from the Animation object, and marks
+the widget tree dirty as necessary, so you don't need to call
+`addListener()`.
+
+The widget tree for this example looks like this:
+
+<img src="images/AnimatedBuilder-WidgetTree.png" alt="a widget tree with Container pointing to ContainerTransition, pointing to AnimatedBuilder, pointing to (AnonymousBuilder), pointing to LogoWidget">
+
+Starting from the bottom of the widget tree, the code for rendering
+the logo is straightforward:
+
+<!-- skip -->
+{% prettify dart %}
+class LogoWidget extends StatelessWidget {
+  // Leave out the height and width so it fills the animating parent
+  build(BuildContext context) {
+    return new Container(
+      margin: new EdgeInsets.symmetric(vertical: 10.0),
+      child: new FlutterLogo(),
+    );
+  }
+}
+{% endprettify %}
+
+The middle three blocks in the diagram are all created in the
```

Comment:
I'm a little confused by the "ContainerTransition" in the diagram. It's not in the code. Is it the same thing as GrowTransition?